sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -586,6 +588,62 @@ object NullPropagation extends Rule[LogicalPlan] {
 }
 
 /**
+ * Attempts to eliminate reading (unnecessary) NULL values if they are not required for correctness
+ * by inserting isNotNull filters is the query plan. These filters are currently inserted beneath
+ * existing Filters and Join operators and are inferred based on their data constraints.
+ *
+ * Note: While this optimization is applicable to all types of join, it primarily benefits Inner and
+ * LeftSemi joins.
+ */
+object NullFiltering extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case filter @ Filter(condition, child: LogicalPlan) =>
+      val isNotNullConstraints = filter.constraints.filter(_.isInstanceOf[IsNotNull])
+      val newCondition = if (isNotNullConstraints.nonEmpty) {
+        And(isNotNullConstraints.reduce(And), condition)
+      } else {
+        condition
+      }
+      Filter(newCondition, child)
+
+    case join @ Join(left: LogicalPlan, right: LogicalPlan, joinType: JoinType,
+      condition: Option[Expression]) =>
+      val leftIsNotNullConstraints = join.constraints
+        .filter(_.isInstanceOf[IsNotNull])
+        .filter(_.references.subsetOf(left.outputSet))
+      val rightIsNotNullConstraints =
+        join.constraints
+          .filter(_.isInstanceOf[IsNotNull])
+          .filter(_.references.subsetOf(right.outputSet))
+      val newLeftChild = if (leftIsNotNullConstraints.nonEmpty) {
+        Filter(leftIsNotNullConstraints.reduce(And), left)
+      } else {
+        left
+      }
+      val newRightChild = if (rightIsNotNullConstraints.nonEmpty) {
+        Filter(rightIsNotNullConstraints.reduce(And), right)
+      } else {
+        right
+      }
+      Join(newLeftChild, newRightChild, joinType, condition)
+  }
+}
+
+/**
+ * Attempts to re-order the individual conjunctive predicates in a filter condition to short circuit
+ * executing relatively cheaper checks (e.g., nullability) before others.
+ */
+object ReorderFilterPredicates extends Rule[LogicalPlan] with PredicateHelper {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case filter @ Filter(condition, child: LogicalPlan) =>
+      val reorderedCondition = splitConjunctivePredicates(condition)
+        .sortWith((x, _) => x.isInstanceOf[IsNotNull])
+        .reduce(And)
+      Filter(reorderedCondition, child)
+  }
+}
```

Comment:
I am wondering if the optimizer is the right place for this rule. My main concern is that if we can preserve this ordering through the rest of query compilation. Will it be better to do it inside the physical Filter operator (just before we start to generate the code)?